pyk/src/pyk/kastManip.py

```diff
@@ -179,6 +160,45 @@ def extract_rhs(term: KInner) -> KInner:
     return top_down(if_ktype(KRewrite, lambda rw: rw.rhs), term)
 
 
+def extract_subst(term: KInner) -> Tuple[Subst, KInner]:
+
+    def _subst_for_terms(term1: KInner, term2: KInner) -> Optional[Subst]:
+        if type(term1) is KVariable and type(term2) not in {KToken, KVariable}:
+            return Subst({term1.name: term2})
+        if type(term2) is KVariable and type(term1) not in {KToken, KVariable}:
+            return Subst({term2.name: term1})
+        return None
+
+    def _extract_subst(conjunct: KInner) -> Optional[Subst]:
+        if type(conjunct) is KApply:
+            if conjunct.label == '#Equals':
+                subst = _subst_for_terms(conjunct.args[0], conjunct.args[1])
+
+                if subst is not None:
+                    return subst
+
+                if conjunct.args[0] == boolToken(True) and type(conjunct.args[1]) is KApply and conjunct.args[1].label in {'_==K_', '_==Int_'}:
+                    subst = _subst_for_terms(conjunct.args[1].args[0], conjunct.args[1].args[1])
+
+                    if subst is not None:
+                        return subst
+
+        return None
+
+    conjuncts = flattenLabel('#And', term)
+    subst = Subst()
+    rem_conjuncts: List[KInner] = []
+
+    for conjunct in conjuncts:
+        new_subst = _extract_subst(conjunct)
+        if new_subst is None:
+            rem_conjuncts.append(conjunct)
+        else:
+            subst = subst.compose(new_subst)
```

Comment:
I guess this should be `union`. How should we handle the case when `subst.union(new_subst) is None`?

Reply:
I think it should be a fatal error basically. We shouldn't have two different bindings for the same variable in the constraints.

It could very well be sound logically, but for now let's just make it a fatal error and then we'll be forced to deal with that case correctly when it comes up.

Reply:
Also, maybe this can be changed to say `if not new_subst:` instead of `if new_subst is None:`? I'm not sure when to use one or the other.

Reply:
In this particular case it is safe to do `not new_subst`, as `extract_subst` never returns `Subst()`. But I'd like to keep it like this so that the reader doesn't have to think through this cornercase.

Reply:
> So you can have a check at the end of the function if not subst: fatal('Producing subst failed!')

Done.